src/core/trulens/core/experimental/__init__.py

```diff
@@ -0,0 +1,441 @@
+        """Determine if the given setting is locked."""
+
+        return self.settings[Feature(flag)].is_locked
+
+    def set(
+        self,
+        flag: Union[Feature, str],
+        *,
+        value: Optional[bool] = None,
+        lock: bool = False,
+    ) -> bool:
+        """Get/Set the given feature flag to the given value.
+
+        Sets the flag to the given value if the value parameter is set. Locks
+        the flag if the lock parameter is set to True.
+
+        Raises:
+            ValueError: If the flag was already locked to a different value.
+        """
+
+        return self.settings[Feature(flag)].set(value=value, lock=lock)
+
+    def lock(
+        self, flag: Union[Feature, str], *, value: Optional[bool] = None
+    ) -> bool:
+        """Lock the given feature flag to the given value.
+
+        If the value is not provided, lock the flag to its current value.
+
+        Raises:
+            ValueError: If the flag has already been locked to a different value.
+        """
+
+        return self.set(flag, value=value, lock=True)
+
+    def get(self, flag: Union[str, Feature], *, lock: bool = False) -> bool:
+        """Determine the value of the given feature flag by checking both global
+        and instance flags.
+
+        Instance value takes precedence over the global value.
+        """
+
+        return self.set(flag, lock=lock)
+
+    def enable(self, flag: Union[Feature, str], *, lock: bool = False) -> bool:
+        """Enable the given feature flag.
+
+        Locks the flag if the lock parameter is set to True.
+
+        Raises:
+            ValueError: If the flag was already locked to disabled.
+        """
+
+        return self.set(flag, value=True, lock=lock)
+
+    def disable(self, flag: Union[Feature, str], *, lock: bool = False) -> bool:
+        """Disable the given feature flag.
+
+        Locks the flag if the lock parameter is set to True.
+
+        Raises:
+            ValueError: If the flag was already locked to enabled.
+        """
+
+        return self.set(flag, value=False, lock=lock)
+
+    def set_multiple(
+        self,
+        flags: Union[
+            Iterable[Union[str, Feature]],
+            Mapping[Union[str, Feature], bool],
+        ],
+        lock: bool = False,
+    ):
+        """Set multiple feature flags.
+
+        If lock is set, lock the flags. If a dictionary is passed, the keys are
+        the feature flags and the values are the values to set them to. If a
+        list is passed, the flags are set to True.
+
+        Raises:
+            ValueError: If any of the flags are already locked to a different
+                value than specified.
+        """
+
+        if isinstance(flags, dict):
+            for flag, val in flags.items():
+                self.set(flag, value=val, lock=lock)
+        else:
+            for flag in flags:
+                self.set(flag, value=True, lock=lock)
+
+
+class WithExperimentalSettings(
```

Comment:
difference between this `WithExperimentalSettings` and `Feature` classes is unclear to me. Can you explain?

Reply:
You mean difference between `WithExperimentalSettings` and `ExperimentalSettings`? Here are all the classes:

- `Feature` -- enum for experimental features, public.
- `Setting` (now `_Setting`) -- class to manage the value to assign to some `Feature` (a boolean True for enabled, False for disabled).
- `ExperimentalSettings` (now `_ExperimentalSettings`) -- class to manage assignments to multiple features.
- `WithExperimentalSettings` -- a mixin class to include in another class to give it capabilities of `ExperimentalSettings` class while printout out messages to the user.

Without a mixin, we could have added the methods in it directly into `TruSession` but that would make a large complex class already more complex. Instead we "mixin" `WithExperimentalSettings`.

Alternatively we could have not had a mixin class at all, and used just `ExperimentalSettings` and try to mix that into TruSession instead. The problem there would be collisions with attributes/names. I want to only include methods that start with `experimental_` into `TruSession` and wanted to seperate the user printouts side of things from the feature value logic. So the solution in the PR has a separate `ExperimentalSettings` which I will rename to `_Settings` which does not print anything and has its own namespace, and a mixin class `WithExperimentalSettings` that provides access to `_Settings` through specialized methods that have user facing printouts.